tensorboard/lazy.py

```diff
@@ -19,36 +19,55 @@
 from __future__ import division
 from __future__ import print_function
 
-import importlib
+import functools
 import types
 
 
 class LazyLoader(types.ModuleType):
-  """Lazily import a module, mainly to avoid pulling in large dependencies."""
+  """Lazily import a module.
+
+  This can be used to defer importing troublesome dependencies - e.g. ones that
+  are large and infrequently used, or that cause a dependency cycle -
+  until they are actually used.
+  """
 
   # The lint error here is incorrect.
-  def __init__(self, local_name, parent_module_globals, name):  # pylint: disable=super-on-old-class
-    self._local_name = local_name
-    self._parent_module_globals = parent_module_globals
+  def __init__(self, name, load_fn):  # pylint: disable=super-on-old-class
+    """Create a LazyLoader for a module.
 
+    Args:
+      name: the fully-qualified name of the module
+      load_fn: callable that actually does the import and returns the module
+    """
     super(LazyLoader, self).__init__(name)
+    self._load_fn = load_fn
```

Comment:
If I understand correctly, this does have the effect that `_load_fn` and
`_cached_module` leak onto the actual module being imported, and in the
event that the underlying module actually defines those properties they
will not be accessible (because of the `__getattr__`/`__getattribute__`
distinction). This might sound like “who would ever do that”, but it can
happen if you try to lazily load a module that is itself lazily loaded:

```python
>>> from tensorboard import lazy
>>> @lazy.lazy_load("inner")
... def inner():
...   import sys
...   return sys
... 
>>> @lazy.lazy_load("outer")
... def outer():
...   return inner
... 
>>> outer.stdout
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "tensorboard/lazy.py", line 55, in __getattr__
    return getattr(self._cached_module, item)
AttributeError: 'NoneType' object has no attribute 'stdout'
>>> inner.stdout
<open file '<stdout>', mode 'w' at 0x7fa4487f5150>
>>> outer.stdout
<open file '<stdout>', mode 'w' at 0x7fa4487f5150>
>>> 
```

This seems a bit insidious. Can we fix the leaky abstraction? For
instance, we could store `_load_fn` and `_cached_module` in a closure
value instead of on the object itself.

Reply:
Noted offline that this is not new in this PR, so it’s fine with me if
you don’t address it and you or I fixes it in a subsequent commit.

Reply:
Refactored to use a closure.